### PR TITLE
Navigation functionality

### PR DESCRIPTION
**Pull request by nirvn**
The time has come: this PR implements basic navigation functionalities to QField :partying_face: 

The navigation mode can be kick-started through a number of ways:
- when viewing a feature, the feature menu has a new action to set the feature as destination (for non-point features, a point on the line or an area centroid is used as destination)
image

- a coordinate can be entered directly into the search bar, where a new set as destination action (pictured using a red flag) is now present
image

- similarly, when looking for features through the search bar, the same set as destination red flag action is not attached to matching features
image

- finally, when long tapping on the map canvas, a brand new menu pops up with a useful pair of X/Y (or lon./lat.) items shows up as well as a "Set as destination" action
image

When navigation is active (i.e., when a destination is present), QField gains a few relevant UI elements.
image

As seen in the screenshot above, a new navigation button appears on top of the location button. When users tap the button, the canvas will automatically pan to recenter the map around the destination point. When long pressing the button, a menu will appear which allows users to clear the navigation destination.

At the bottom of the screen, a new navigation information panel appears with details such as destination coordinate, distance to destination, and bearing to destination.

When both a destination and positioning is turned on, a line is drawn on the map canvas to indication the direction to/from the current location and destination:
Peek 2022-02-26 17-07

When the destination marker (a red diamond) isn't visible within the current map canvas extent, an edge indicator is present to give users a sense of where the marker is. A similar edge indicator was also implemented for the current location. At the moment, the indicator is quite subtle, I'm open to feedback on whether they should be bigger.

### CODE REVIEW

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 8b575085b963efe32f8d4e44d0b9044064450390:
  - [**arm64-android**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2539-375e311eda6dcc52a30d1469cbd30d745088ce9f-arm64-android.apk)

Other architectures: [arm-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2539-375e311eda6dcc52a30d1469cbd30d745088ce9f-arm-android.apk), [x64-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2539-375e311eda6dcc52a30d1469cbd30d745088ce9f-x64-android.apk), [x86-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2539-375e311eda6dcc52a30d1469cbd30d745088ce9f-x86-android.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for bc979c26f1789351b9de280b1af84999048b4e66:
  - [**arm64-android**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2539-7c01696d735098bf28c80fc90280397a1c85f486-arm64-android.apk)

Other architectures: [arm-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2539-7c01696d735098bf28c80fc90280397a1c85f486-arm-android.apk), [x64-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2539-7c01696d735098bf28c80fc90280397a1c85f486-x64-android.apk), [x86-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2539-7c01696d735098bf28c80fc90280397a1c85f486-x86-android.apk)

**Comment by nirvn** [comment]
Forgot to mention one implementation detail: the navigation destination is remembered across sessions.

File: src/qml/LocatorItem.qml [hunk 1]
```diff
@@ -66,7 +66,6 @@ Item {
     keepScale: qfieldSettings.locatorKeepScale
 
     featureListController: featureForm.extentController
-    bookmarks: bookmarkModel
```
Comment (suricactus) [inline src/qml/LocatorItem.qml:69]: Is this on purpose?

File: src/qml/NavigationHighlight.qml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  property Navigation navigation
+
+  LinePolygonHighlight {
+    visible: positionSource.active
+    mapSettings: navigation.mapSettings
+    geometry:   QgsGeometryWrapper {
+      qgsGeometry: navigation.path
+      crs: navigation.mapSettings.crs
+    }
+    color: "#77FF0000"
```
Comment (suricactus) [inline src/qml/NavigationHighlight.qml:19]: Can you add this color to the theme and/or name it?

File: src/qml/NavigationRenderer.qml [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+                        mapSettings: mapCanvas.mapSettings
+                        mapPoint: _ct.projectedPosition
+                    }
+
+                    property bool isOnCanvas: mapToScreenPosition.screenPoint.x > 0
+                                              && mapToScreenPosition.screenPoint.x < mapCanvas.width
+                                              && mapToScreenPosition.screenPoint.y > 0
+                                              && mapToScreenPosition.screenPoint.y < mapCanvas.height
+                    x: Math.min(mapCanvas.width, Math.max(0,mapToScreenPosition.screenPoint.x)) - width/2
+                    y: Math.min(mapCanvas.height, Math.max(0,mapToScreenPosition.screenPoint.y)) - height/2
```
Comment (suricactus) [inline src/qml/NavigationRenderer.qml:61]:
```suggestion
                    x: Math.min(mapCanvas.width, Math.max(0, mapToScreenPosition.screenPoint.x)) - width / 2
                    y: Math.min(mapCanvas.height, Math.max(0, mapToScreenPosition.screenPoint.y)) - height / 2
```
A bit of spacing is nice :)

File: src/qml/qgismobileapp.qml [hunk 4]
```diff
@@ -562,31 +576,38 @@ ApplicationWindow {
...
     anchors.left: parent.left
     anchors.right: parent.right
-    visible: positioningSettings.showPositionInformation
+    visible: navigation.isActive || positioningSettings.showPositionInformation
 
     height: childrenRect.height
     width: parent.width
 
+    NavigationInformationView {
+      id: n
```
Comment (suricactus) [inline src/qml/qgismobileapp.qml:590]: sorry, can we have better ids?

File: src/qml/qgismobileapp.qml [hunk 5]
```diff
@@ -973,6 +997,21 @@ ApplicationWindow {
...
+      visible: navigation.isActive
+      round: true
+
+      iconSource: Theme.getThemeIcon( "ic_navigation_flag_red_24dp" )
+      bgcolor: Theme.darkGray
+
+      onClicked: mapCanvas.mapSettings.setCenter(navigation.destination)
+
+      onPressAndHold: {
+        navigationMenu.popup(locationToolbar.x + locationToolbar.width - navigationMenu.width, locationToolbar.y + locationToolbar.height - navigationMenu.height)
```
Comment (suricactus) [inline src/qml/qgismobileapp.qml:1011]:
```suggestion
        navigationMenu.popup(
          locationToolbar.x + locationToolbar.width - navigationMenu.width,
          locationToolbar.y + locationToolbar.height - navigationMenu.height
        )
```

File: src/qml/qgismobileapp.qml [hunk 6]
```diff
@@ -1613,6 +1652,88 @@ ApplicationWindow {
...
+    title: qsTr( "Map Canvas Options" )
+    font: Theme.defaultFont
+
+    property var point
+    onPointChanged: {
+      xItem.text = ( mapCanvas.mapSettings.destinationCrs.isGeographic ? qsTr( 'Lon' ) : 'X' ) + ': ' +
+                   Number( point.x ).toLocaleString( Qt.locale(), 'f', coordinateLocator.mapSettings.destinationCrs.isGeographic ? 7 : 3 )
+      yItem.text = ( mapCanvas.mapSettings.destinationCrs.isGeographic ? qsTr( 'Lat' ) : 'Y' ) + ': ' +
+                   Number( point.y ).toLocaleString( Qt.locale(), 'f', coordinateLocator.mapSettings.destinationCrs.isGeographic ? 7 : 3 )
+    }
```
Comment (suricactus) [inline src/qml/qgismobileapp.qml:1666]:
```suggestion
    onPointChanged: {
      var xLabel = mapCanvas.mapSettings.destinationCrs.isGeographic ? qsTr( 'Lon' ) : 'X';
      var xValue = Number( point.x ).toLocaleString( Qt.locale(), 'f', coordinateLocator.mapSettings.destinationCrs.isGeographic ? 7 : 3 )
      var yLabel = mapCanvas.mapSettings.destinationCrs.isGeographic ? qsTr( 'Lat' ) : 'Y'
      var yValue = Number( point.y ).toLocaleString( Qt.locale(), 'f', coordinateLocator.mapSettings.destinationCrs.isGeographic ? 7 : 3 )

      xItem.text = xLabel + ': ' + xValue
      yItem.text = yLabel + ': ' + yValue
    }
```

More readable version, took me a while to understand what you did here.

File: src/core/navigationmodel.cpp [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+void NavigationModel::restore()
+{
+  QSettings settings;
+  if ( settings.contains( QStringLiteral( "/QField/navigation/points" ) ) )
+  {
+    beginResetModel();
+    settings.beginGroup( QStringLiteral( "/QField/navigation" ) );
+    mCrs.createFromWkt( settings.value( QStringLiteral( "crs" ), QString() ).toString() );
+    mPoints.clear();
+    const QStringList pointWkts( settings.value( QStringLiteral( "points" ), QString() ).toString().split( QStringLiteral( "}||{" ) ) );
```
Comment (suricactus) [inline src/core/navigationmodel.cpp:167]:
```suggestion
    const QStringList pointWkts( 
    	settings.value( QStringLiteral( "points" ) ).toString().split( QStringLiteral( "}||{" ) ) );
```
Should work fine without a default, improves a tiny bit the readability. Is using "split" the only way to parse the value?

File: src/core/navigationmodel.cpp [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+  {
+    QStringList pointWkts;
+    for ( const QgsPoint &point : mPoints )
+    {
+      pointWkts << point.asWkt();
+    }
+
+    settings.beginGroup( QStringLiteral( "/QField/navigation" ) );
+    settings.setValue( QStringLiteral( "crs" ), mCrs.toWkt() );
+    settings.setValue( QStringLiteral( "points" ), pointWkts.join( QStringLiteral( "}||{" ) ) );
```
Comment (suricactus) [inline src/core/navigationmodel.cpp:149]: Not sure if this is my favourite serialization format.

File: src/qml/LocatorItem.qml (hunk 1 above)
Comment (nirvn) [reply src/qml/LocatorItem.qml:69]: @suricactus , correct

File: src/qml/NavigationHighlight.qml (hunk 2 above)
Comment (nirvn) [reply src/qml/NavigationHighlight.qml:19]: done

File: src/qml/NavigationInformationView.qml [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+Rectangle {
+  id: navigationInformationView
+
+  property Navigation navigation
+  property int ceilsCount: 4
+  property double rowHeight: 30
+  property color backgroundColor: "white"
+  property color alternateBackgroundColor: "#ffd3d3"
+  property color textColor: "black"
```
[the remark this answers is not in the document]
Comment (nirvn) [reply src/qml/NavigationInformationView.qml:18]: Moved the alt. background color into theme and applied the same improvements to the position information panel. The white and black color will stay as such for now, improvements there can come in whenever we add a dark theme variant to QField.

File: src/core/navigationmodel.cpp (hunk 8 above)
Comment (nirvn) [reply src/core/navigationmodel.cpp:149]: OK, just saving as string list now.

**Comment by nirvn** [comment]
@suricactus , thanks for the review, comments addressed (or replied to)

On your additional points in the last comment:
- 0 -> fixed, thanks
- 1 -> based on discussions with people that use high precision GNSS devices, there's a pretty consistent "give us 3 decimals all the time", so I'm inclined to stick with that consistently
- 2 -> @m-kuhn suggested purple as alternative, but I'm worried for colorblind folks, purple navigation and blue location might blend in too much
- 3 -> lots can be done with this new UI/UX, copying the location is a great idea, can do in a follow up PR during free time
- 4 -> I assume you mean the search bar? Maybe we can try to adjust the behavior in a follow up PR, that has implications beyond this navigation implementation

**Comment by nirvn** [comment]
@suricactus , @m-kuhn , OK, turns out purple is indeed a better color pick :)
image

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 4b3fb6d1ba0b8e8a28fe800d7e72466aa3bb0c9b:
  - [**arm64-android**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2539-45b3f8f52d917e58ecf603d050ecb25b90e1b472-arm64-android.apk)

Other architectures: [arm-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2539-45b3f8f52d917e58ecf603d050ecb25b90e1b472-arm-android.apk), [x64-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2539-45b3f8f52d917e58ecf603d050ecb25b90e1b472-x64-android.apk), [x86-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2539-45b3f8f52d917e58ecf603d050ecb25b90e1b472-x86-android.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 590f82e16602a1cfdec9ec15e022421b5d4e6309:
  - [**arm64-android**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2539-c5a7cd839eae6821fbac011f793b073395ec5f01-arm64-android.apk)

Other architectures: [arm-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2539-c5a7cd839eae6821fbac011f793b073395ec5f01-arm-android.apk), [x64-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2539-c5a7cd839eae6821fbac011f793b073395ec5f01-x64-android.apk), [x86-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2539-c5a7cd839eae6821fbac011f793b073395ec5f01-x86-android.apk)

**Comment by m-kuhn** [comment]
I like the 💜
Would it make sense to give it a bit more width?

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 62e9a837e80e6be4b883cbe678a788f24fb8a949:
  - [**arm64-android**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2539-fd9f14ecf4955d4ec7d63357e1766e9b4053adb2-arm64-android.apk)

Other architectures: [arm-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2539-fd9f14ecf4955d4ec7d63357e1766e9b4053adb2-arm-android.apk), [x64-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2539-fd9f14ecf4955d4ec7d63357e1766e9b4053adb2-x64-android.apk), [x86-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2539-fd9f14ecf4955d4ec7d63357e1766e9b4053adb2-x86-android.apk)

**Comment by nirvn** [comment]
Alright, all comments addressed.

I'd like to merge this if you are OK. I have additional tweaks I would like to address outside of a >1,000 line addition PR :)

**Comment by suricactus** [comment]
Much better with the purple!

Didn't check the code again, seems to work for me. Ready to merge!

**Review by suricactus — APPROVED** [review]

**Pull request merged**